README
This documentation is fairly out of date. For the latest, please go to
the project homepage at https://github.com/squash/sudosh2

Introduction:
=============

sudosh is a filter and can be used as a login shell.  sudosh takes advantage
of pty devices in order to sit between the user's keyboard and a program, in
this case a shell.

sudosh was designed specifically to be used in conjunction with sudo or by
itself as a login shell..  sudosh allows the execution of a root shell with
logging.  Every command the user types within the root shell is logged as
well as the output.

How is this different than "sudo -s" or "sudo /bin/sh" ?

Using "sudo -s" or other methods doesn't log commands typed to syslog.
Generally the commands are logged to a file such as .sh_history and if you
use a shell such as csh that doesn't support command-line logging you're
out of luck.

sudosh fills this gap.  No matter what shell you use, all of the command
lines are logged to syslog (including vi keystrokes.)

Installation:
=============

Compile sudosh.

1) ./configure
2) make
3) make install
4) sudosh -i

Configure sudosh to be used with sudo
=====================================

1) configure /etc/sudoers to allow system administrators to execute
   /usr/local/bin/sudosh

Example entry to /etc/sudoers:

-- /etc/sudoers begin --
User_Alias      ADMINS=admin1,admin2,admin3
User_Alias      DBAS=dba1,dba2,dba3
Cmnd_Alias      SUDOSH=/usr/local/bin/sudosh

ADMINS          ALL=SUDOSH
DBAS            ALL=(oracle)/usr/local/bin/sudosh
-- /etc/sudoers end --

2) use it.
3) Look at the results with the sudosh-replay command.

Example usage to get ROOT ACCESS:

-- example usage begin --
# Get access to root
dhanks@linux:~> sudo sudosh
Password:
starting session for dhanks as root,/dev/tty3 (/bin/bash) (root-1108447320)
linux:~ # id
uid=0(root) gid=0(root) groups=0(root)
linux:~ # exit
exit
dhanks@linux:~>
-- example usage end --

Example usage to get ORACLE ACCESS:

-- example usage begin --
# Get access to oracle
dhanks@linux:~> sudo -u oracle sudosh
Password:
starting session for dhanks as oracle,/dev/tty3 (/bin/bash) (oracle-1108447391)
oracle@linux:~> id
uid=1001(oracle) gid=100(users) groups=100(users)
oracle@linux:~> exit
dhanks@linux:~>
-- example usage end --

Configure sudosh to be used as a login shell
============================================

1) Set the default shell to be used as a login shell.  If you do not change
   the default value /bin/sh will be used.  Use the --with-defshell option
   to configure to set this value.

2) If your system supports /etc/shells, add the absolute path to sudosh to
   the list.

3) Edit /etc/passwd and set the user's shell to the absolute path of
   sudosh (generally /usr/local/bin/sudosh unless you change the --prefix)

4) To allow things like scp, sft, cvs, rsync, etc, use the "-c arg allow" in
   sudosh.conf. The allowed command will need to match what is being sent to
   it, notably some OpenSSH servers will call /usr/lib/openssh/sftp-server.

How To Replay Sessions
======================
Use the "sudosh-replay" command to replay previous root sessions.  To see
a list of available sessions execute "sudosh-replay" by itself.

-- example begin --
linux:~ # sudosh-replay
Sessions stored in /var/log/sudosh:
===================================
dhanks       1 session
  Wed Oct 20 18:39:00 2004 (ID dhanks-1098322740)

Usage: sudosh-replay ID [MULTIPLIER] [MAXWAIT]
Example: sudosh-replay dhanks-1098322740 1 2
linux:~ #
-- example end--

As we can see there is 1 session available to view on my system at the
time of writing this documentation.

Use the ID to specify which session you which to replay.  The ID is located
in the brackets after the date string.  In this case our ID is
"dhanks-1098322740"

To view this session just as the user typed it, execute
"sudosh-replay dhanks-1098322740"

To merely just output the session data and ignore the timing information,
just set the multiplier to 0:
"sudosh-replay dhanks-1098322740" 0

The multiplier also works to speed up the session.  If you wish to speed up
the session 2x, use a multiplier of 2

"sudosh-replay dhanks-1098322740" 2

If you have people that type slowly or if a user goes away from the keyboard,
you don't want to sit there and wait for the output.  Instead set the MAXWAIT
variable.  It's set to 1 by default.

"sudosh-replay dhanks-1098322740" 2 .5


NEWS
NEWS file is outdated.

April 22 2015:
	Moved to Github. 

December 20 2007:
	Created new Sourceforge project for sudosh2. This is
	a continuation of development of sudosh, which
	appears to be abandoned.

	I will be releasing 1.0 via sourceforge today with a
	couple of patches required in my environment (RHEL
	3, 4, and 5, both 32 and 64 bit, and Solaris 9 Sparc)

	Bugs fixed include a config file parsing bug that would
	lead to a segfault, and a fix for pwent handling which
	would cause issues with old vs new uids.

	I also removed the requirement to initialize your log 
	directory. It ignores the -i option, which is left in
	for compatibility with old install scripts. sudosh will
	check the logdir for existance and correct permissions
	every time it is run, and create or correct as necessary.

June 12th 2005:
	Added configuration file support to sudosh.  You can now
	setup:
		logdir
		default shell
		delimiter
		syslog.priority
		syslog.facility
		-c arg allow

	Also added -c support to sudosh so that remote commands
	can be executed.  This allows scp and rsync to be executed
	remotely.

	I've also ported sudosh-replay to C.

	I decided to stop supporting old versions of sudosh.  Use
	the newest release instead.  Anyone using 1.4.3 definately
	needs to upgrade.  The new version of sudosh only works with
	newer logfiles.

May 12th 2005:
	Silly bug in sudosh-replay wasn't allowing the new file
	delimiter option to function completely.

	Removed the dependacy from the TERM environment variable.
	If it cannot be determined from the existing environment
	it will default to the setting of "dumb"  This way things
	such as scp, rsync and the like will function properly if
	sudosh is to be used as a login shell.

May 11th 2005:
	Finally gave into all the requests for sudosh to be used as
	a login shell from /etc/passwd.  I'm not sure why people want
	to do this.  Sudosh should be used during esclation.  For
	example a user should login to UNIX with a personal account
	that has minimal privileges.  From here sudo(8) should be used
	to gain access to other privileges such as oracle or other
	application accounts.  This is where sudosh is to be used,
	for example:

	1) login with personal account 'dhanks'
	2) use sudo to gain access to oracle with sudosh.
		$ sudo -u oracle sudosh
	3) use the oracle shell with logging.

	Regardless, I've added the ability to use sudosh as a login
	shell.  Just be sure to add the absolute path to sudosh in
	/etc/shells and in /etc/passwd for the desired user.  The
	default shell is defined in configure as /bin/sh.  If you
	need to change it do so with the --with-defshell option to
	configure, make, make install.

	Luke Youngblood emailed me the other day regarding the file
	delimiter of files used with sudosh.  Generally the files are
	delimited with a dash '-'.  Luke pointed out he uses a dash
	in the usernames such as "username-prod"  Obviously this was
	a problem for sudosh because it relies on the dash to know
	which files are which.

	Use the --with-filedelimit option to configure the default
	file delimiter.  I've successfully tested a dash '-', the
	underline '_', and the colon ':'.

May 9th 2005:
	Had problems with AIX.  I love AIX, but forgot AIX has special
	subroutines that handle user attributes very nicely.  I've added
	support for setpenv() which takes care of setting a clean
	environment, chdir() to user's home directory and exec()ing the
	user's shell all defined from /etc/passwd and /etc/environment.

	Added MAXWAIT support to sudosh-replay.  Thanks to Luc Charland
	for the patch.  This was something that always bothered me about
	sudosh-replay, but I never got around to adding support for a
	MAXWAIT.  Added the usage to the README.

	Had to make a small correction on the chdir() because accidentally
	used the original HOME instead of the "to" user's home.

	I'm still struggling with execv() vs. execle().  execle() seems
	to be working fine for now.  Working with Todd on this one.

	I've certified sudosh 1.6.1 on AIX, Solaris and Linux.  I've had
	user reports that it works fine on HPUX and Tru64.  I used the
	command:

	PATH="" sudosh

	Then I proceded to check the environment.  PATH was set correctly
	by the shell through /etc/profile and the like.  On AIX it was
	set through setpenv() via /etc/environment.

May 7th 2005:
        Sudosh now calls shells as if they were a login shell.  Sudosh
        automatically sets the environment variables LOGNAME, USER, TERM,
        SHELL and PATH.  Changed the way the shell is executed.  Added a
	dash '-' before the arguement, i.e. -ksh  This allows the shell
	to execute as if it were a login shell.

        Added a random string to the log files so that non-privileged users
        cannot delete their log files.  Removed tampering code because it
	was not portable.  General code cleanup.

April 12th 2005:
	Added functions to monitor log files and alert to syslog when
	someone attempts to remove or modify the files.

March 22nd 2005:
	Cleaned up the sudosh environment.  TERM wasn't being transferred.

March 11th 2005:
	Corrected incorrect HOME environment variable.

March 10th 2005:
	Added setenv("HOME", pw->pw_dir, 1); so that sudosh will know
	where the user's home directory is.  This is helpful when the user's
	shell is exec()'d and it needs to find the .profile and other shell
	files.

March 8th 2005:
	Added man pages for sudosh and sudosh-replay.

March 5th 2005:
	Changed the way sudosh handles LOGDIR permissions.  I had a
	problem with st_mode == logdir because through AIX and Solaris
	there are some extra higher level flags that throw it off.  For
	now I will ignore the validation level until the next release.

	Chris MacLeod submited two patches for sudosh.  Thank you very
	much.  Chris contributed:

		1) specfile for building on redhat.

		2) profile.d script for bash.  Adds a sudo function so
		   that 'sudo -s' or just 'sudo' works seemlsesly.

February 19th 2005:

src/replay.c
#include "super.h"
#include "struct.h"

#define LL() fprintf(stderr, "[%s, line %i]: ", __FILE__, __LINE__)

struct s_file
{
  FILE *f;
  int fd;
  int line;
  char *ptr;
  char str[BUFSIZ];
};

struct s_session
{
  char to[BUFSIZ];
  char from[BUFSIZ];
  char type[BUFSIZ];
  char randstr[BUFSIZ];
  char date[BUFSIZ];
  char id[BUFSIZ];
  long secs;
  time_t e;
  struct s_session *next;
  struct s_session *prev;
  struct s_file script;
  struct s_file time;
};

typedef struct s_session session;

extern void parse (option *, const char *);

int getopt (int, char *const[], const char *);
extern char *optarg;
extern int optind;

char match[BUFSIZ];

char *progname;
session *s_top;
option config_option;

void replay (const char *, const char *, int, int);
void link_session (session *);
void unlink_session (session *);
void count_dur (session *);
char *int2smalltime (int);

session *session_malloc (void);
session *find_session (const char *, const char *, time_t e, const char *);
session *sort_list (session *);
void show_sessions (void);

int
main (int argc, char **argv, char **environ)
{
  int c;
  char *p = (char *) 0;
  char sysconfdir[BUFSIZ];
  struct dirent *dirp;
  DIR *dp;
  session *scan;
  session *last;

  progname = argv[0];

  if ((p = (char *) strrchr (progname, '/')) != (char *) 0)
    progname = p + 1;


  snprintf (sysconfdir, BUFSIZ - 1, "%s/sudosh.conf", SYSCONFDIR);
  parse (&config_option, sysconfdir);

  snprintf (match, BUFSIZ - 1,
	    "%%32[^%c]%c%%32[^%c]%c%%32[^%c]%c%%ld%c%%64[A-Za-z0-9]",
	    config_option.fdl, config_option.fdl, config_option.fdl,
	    config_option.fdl, config_option.fdl, config_option.fdl,
	    config_option.fdl);

  while (1)
    {
      c = getopt (argc, argv, "d:hVv");

      if (c == EOF)
	break;

      switch (c)
	{
	case 'V':
	case 'v':
	  fprintf (stdout, "%s version %s\n", progname, VERSION);
	  exit (EXIT_SUCCESS);
	  break;
	case 'h':
	  fprintf (stdout, "Usage: %s ID [MULTIPLIER] [MAXWAIT]\n", progname);
	  fprintf (stdout,
		   "Replay %s sessions as they were originally recorded.\n\n",
		   progname);
	  fprintf (stdout,
		   "-d	Specify alternate LOGDIR where sessions are stored.\n");
	  fprintf (stdout, "-h	Show help synopsis.\n");
	  fprintf (stdout, "-v,-V	Show version.\n\n");
	  fprintf (stdout,
		   "ID\n	The session id of which you wish to view.  The format is user-id\n");
	  fprintf (stdout,
		   "	You may obtain a list of ids by just running %s\n\n",
		   progname);
	  fprintf (stdout, "MULTIPLIER\n");
	  fprintf (stdout,
		   "	This is optional.  If you wish to \"speed up\" the playback of the\n");
	  fprintf (stdout,
		   "	session, specify a rate to increase it.  For example if you used\n");
	  fprintf (stdout,
		   "	a multiplier of \"2\" it would playback the session 2x the original\n");
	  fprintf (stdout, "	recording speed.\n\n");
	  fprintf (stdout, "MAXWAIT\n");
	  fprintf (stdout,
		   "	This is optional.  If you do not want to wait during any long pauses\n");
	  fprintf (stdout,
		   "	use this option to specify a maximum wait time.  For example a MAXWAIT\n");
	  fprintf (stdout,
		   "	of 2 would wait a maximum of 2 seconds then continue.\n\n");
	  fprintf (stdout,
		   "Example: how to get a list of sessions to replay.\n");
	  fprintf (stdout, "# %s\n\n", progname);
	  fprintf (stdout,
		   "Example: how to replay a session as it was recorded.\n");
	  fprintf (stdout, "# %s ID\n", progname);
	  fprintf (stdout,
		   "Example: how to replay a session, but just echo the contains to the screen.\n");
	  fprintf (stdout, "# %s ID 0\n\n", progname);
	  fprintf (stdout, "This is %s version %s\n", progname, VERSION);
	  fprintf (stdout, "Report bugs to <%s>\n", PACKAGE_BUGREPORT);
	  exit (EXIT_SUCCESS);
	  break;
	case 'd':
	  strncpy (config_option.logdir, optarg, BUFSIZ - 1);
	  break;
	default:
	  fprintf (stderr, "%s: unrecognized option `%c'\n", progname, c);
	  fprintf (stderr, "Try `%s -h' for more information.\n", progname);
	  exit (EXIT_FAILURE);
	  break;
	}
    }


  argc -= optind;
  argv += optind;

  if ((dp = opendir (config_option.logdir)) == (DIR *) 0)
    {
      LL ();
      fprintf (stderr, "%s: can't opendir(%s): %s (%i)\n", progname,
	       config_option.logdir, strerror (errno), errno);
      exit (EXIT_FAILURE);
    }

  while ((dirp = readdir (dp)) != (struct dirent *) 0)
    {
      session *s;
      char from[BUFSIZ];
      char to[BUFSIZ];
      char type[BUFSIZ];
      char randstr[BUFSIZ];
      time_t e;

      if (strcmp (dirp->d_name, ".") == 0 || strcmp (dirp->d_name, "..") == 0)
	continue;


      if (sscanf (dirp->d_name, match, from, to, type, &e, randstr) != 5)
	{
	  LL ();
	  fprintf (stderr, "[warning]: invalid file: %s\n", dirp->d_name);
	  continue;
	}

      if (strcmp (type, "script") && strcmp (type, "time")
	  && strcmp (type, "interactive") && strcmp (type, "input"))
	{
	  fprintf (stderr, "[warning]: invalid type (%s): %s\n", type,
		   dirp->d_name);
	  continue;
	}

      if (!strcmp (type, "interactive"))
	{
	  char *ptr;
	  FILE *f;
	  s = session_malloc ();

	  s->secs = 0;
	  s->e = e;
	  snprintf (s->script.str, BUFSIZ - 1, "%s/%s", config_option.logdir,
		    dirp->d_name);

	  f = fopen (s->script.str, "r");

	  if (f == (FILE *) 0)
	    {
	      fprintf (stderr, "%.100s: %.100s (%i)\n", s->script.str,
		       strerror (errno), errno);
	      exit (EXIT_FAILURE);
	    }

	  if (fgets (s->id, BUFSIZ - 1, f) == NULL)
	    fprintf (stderr, "Bad header in %s\n", s->script.str);
	  else
	    {
	      ptr = strchr (s->id, '\n');
	      if (ptr != (char *) 0)
		*ptr++ = '\0';

	      strncpy (s->from, from, BUFSIZ - 1);
	      strncpy (s->to, to, BUFSIZ - 1);
	      strncpy (s->type, type, BUFSIZ - 1);
	      strncpy (s->randstr, randstr, BUFSIZ - 1);
	      strftime (s->date, 20, "%m/%d/%Y %H:%M:%S", localtime (&s->e));
	      link_session (s);
	    }
	  fclose (f);
	  continue;
	}

      s = find_session (from, to, e, randstr);

      if (s == (session *) 0)
	{
	  s = session_malloc ();
	  s->secs = 0;

	  if (!strcmp (type, "script"))
	    snprintf (s->script.str, BUFSIZ - 1, "%s/%s",
		      config_option.logdir, dirp->d_name);
	  if (!strcmp (type, "time"))
	    {
	      snprintf (s->time.str, BUFSIZ - 1, "%s/%s",
			config_option.logdir, dirp->d_name);
	      count_dur (s);
	    }

	  s->e = e;
	  strncpy (s->from, from, BUFSIZ - 1);
	  strncpy (s->to, to, BUFSIZ - 1);
	  strncpy (s->type, type, BUFSIZ - 1);
	  strncpy (s->randstr, randstr, BUFSIZ - 1);
	  strftime (s->date, 20, "%m/%d/%Y %H:%M:%S", localtime (&s->e));
	  snprintf (s->id, BUFSIZ - 1, "%s%c%s%c%ld%c%s", s->from,
		    config_option.fdl, s->to, config_option.fdl, s->e,
		    config_option.fdl, s->randstr);

	  link_session (s);
	}
      else
	{
	  if (!strcmp (type, "script"))
	    snprintf (s->script.str, BUFSIZ - 1, "%s/%s",
		      config_option.logdir, dirp->d_name);
	  if (!strcmp (type, "time"))
	    {
	      snprintf (s->time.str, BUFSIZ - 1, "%s/%s",
			config_option.logdir, dirp->d_name);
	      count_dur (s);
	    }
	}
    }

  if (closedir (dp) < 0)
    {
      LL ();
      fprintf (stderr,
	       "%s: can't closedir(%s): %s (%i) .. But teally this isn't fatal, it just wasn't expected.\n",
	       progname, config_option.logdir, strerror (errno), errno);
//      exit(EXIT_FAILURE);
    }

  if (argc)
    {
      char from[BUFSIZ];
      char to[BUFSIZ];
      char randstr[BUFSIZ];
      time_t e;
      int div, maxwait;
      session *s;

      snprintf (match, BUFSIZ - 1,
		"%%32[^%c]%c%%32[^%c]%c%%ld%c%%64[A-Za-z0-9]",
		config_option.fdl, config_option.fdl, config_option.fdl,
		config_option.fdl, config_option.fdl);

      if (sscanf (argv[0], match, from, to, &e, randstr) != 4)
	{
	  LL ();
	  fprintf (stderr, "%s: invalid format: %s\n", progname, argv[0]);
	  fprintf (stderr, "%s: the appropiate format is %s\n", progname,
		   match);
	  exit (EXIT_FAILURE);
	}

      s = find_session (from, to, e, randstr);

      if (s == (session *) 0)
	{
	  LL ();
	  fprintf (stderr, "%s: no such session: %s\n", progname, argv[0]);
	  exit (EXIT_FAILURE);
	}

      div = 1;
      maxwait = 1;

      if (argc > 1)
	div = atoi (argv[1]);
      if (argc > 2)
	maxwait = atoi (argv[2]);

      if (div == 0)
	{
	  FILE *f_script = fopen (s->script.str, "r");
	  char buffer[BUFSIZ];

	  if (f_script == (FILE *) 0)
	    {
	      LL ();
	      fprintf (stderr, "%s: %s: %s (%i)\n", progname,
		       s->script.str, strerror (errno), errno);
	      exit (EXIT_FAILURE);
	    }

	  while (fread (&buffer, 1, BUFSIZ, f_script))
	    fputs (buffer, stdout);

	  fclose (f_script);

	}
      else
	replay (s->time.str, s->script.str, div, maxwait);

      exit (EXIT_SUCCESS);
    }

  if (!s_top)
    {
      fprintf (stderr, "No sessions are logged in %s\n",
	       config_option.logdir);
      exit (EXIT_SUCCESS);
    }

  fprintf (stderr, "%-19s %-8s %-12s %-12s %s\n", "Date", "Duration",
	   "From", "To", "ID");
  fprintf (stderr, "%-19s %-8s %-12s %-12s %s\n", "====", "========",
	   "====", "==", "==");

  s_top = sort_list (s_top);
  last = s_top;

  for (scan = s_top; scan; scan = scan->next)
    {
      if ((scan->script.fd = open (scan->script.str, O_RDONLY)) == -1)
	{
	  LL ();
	  fprintf (stderr,
		   "[error]: session %s is missing script information.\n",
		   scan->id);
	  continue;
	}
      close (scan->script.fd);

      if (!strcmp (scan->type, "interactive"))
	fprintf (stdout, "%-19s %-8s %-12s %-12s %s\n", scan->date,
		 "-c arg", scan->from, scan->to, scan->id);
      else
	fprintf (stdout, "%-19s %-8s %-12s %-12s %s\n", scan->date,
		 int2smalltime (scan->secs), scan->from, scan->to, scan->id);
      last = scan;
    }

  if (s_top && last)
    fprintf (stderr, "\nUsage: %s ID [MULTIPLIER] [MAXWAIT]\n"
	     "See '%s -h' for more help.\n"
	     "Example: %s %s 1 2\n", progname, progname, progname, last->id);
  else
    fprintf (stderr, "\nUsage: %s ID [MULTIPLIER] [MAXWAIT]\n"
	     "See '%s -h' for more help.\n", progname, progname);

  exit (EXIT_SUCCESS);
}

void
replay (const char *time, const char *script, int div, int maxwait)
{
  char read_buffer[BUFSIZ];
  char timebuf[BUFSIZ];
  float ftime = 0;
  int b = 0;
  int r = 0;
  int sec, usec;
  char buffer[BUFSIZ];
  struct s_file s_time;
  struct s_file s_script;
  struct timeval tv;

  if (div < 1)
    div = 1;

  if (maxwait < 1)
    maxwait = 1;

  s_time.f = fopen (time, "r");

  if (s_time.f == (FILE *) 0)
    {
      LL ();
      fprintf (stderr, "%s: %s: %s (%i)\n", progname, time,
	       strerror (errno), errno);
      exit (EXIT_FAILURE);
    }

  if ((s_script.fd = open (script, O_RDONLY)) == -1)
    {
      LL ();
      fprintf (stderr, "%s: %s: %s (%i)\n", progname, script,
	       strerror (errno), errno);
      exit (EXIT_FAILURE);
    }

  for (s_time.line = 1; fgets (buffer, BUFSIZ - 1, s_time.f); s_time.line++)
    {

      if (sscanf (buffer, "%s %i\n", (char *) &timebuf, &b) != 2)
	{			/* find the two numbers */
	  LL ();
	  fprintf (stderr,
		   "[error]: line %i: invalid timing information (!= %%f %%i\\n)\n",
		   s_time.line);
	  exit (EXIT_FAILURE);
	}

      if (sscanf (timebuf, "%f", &ftime) != 1)
	{
	  LL ();
	  fprintf (stderr, "[error]: line %i: invalid float (!= %%f)",
		   s_time.line);
	  exit (EXIT_FAILURE);
	}

      if (ftime / div > 0.0001)
	if (ftime / div > maxwait)
	  ftime = maxwait / div;
      ftime = ftime / div;

      snprintf (timebuf, BUFSIZ - 1, "%f", ftime);

      if (sscanf (timebuf, "%i.%i", &sec, &usec) != 2)
	{
	  LL ();
	  fprintf (stderr, "[error]: line %i: invalid float (!= %%i.%%i)",
		   s_time.line);
	  exit (EXIT_FAILURE);
	}

      tv.tv_sec = (time_t) sec;
      tv.tv_usec = (time_t) usec;

      if (b > 1024 * 1024 * 8)
	{			/* 8MB */
	  LL ();
	  fprintf (stderr,
		   "[error]: line %i: wanted to read %i bytes, but the limit is 8MB.\n",
		   s_time.line, b);
	  exit (EXIT_FAILURE);
	}

      memset (read_buffer, '\0', BUFSIZ);
      r = read (s_script.fd, read_buffer, (size_t) b);

      if (r != b)
	{
	  LL ();
	  fprintf (stderr, "[failure]: read %i bytes out of %i.\n", r, b);
	  exit (EXIT_FAILURE);
	}

      select (0, (fd_set *) 0, (fd_set *) 0, (fd_set *) 0, &tv);
      fputs (read_buffer, stdout);
      fflush (stdout);
      memset (read_buffer, '\0', BUFSIZ);
    }
  close (s_script.fd);
  fclose (s_time.f);
  fprintf (stderr, "[info]: EOF\n");
  fflush (stderr);
}

session *
session_malloc (void)
{
  session *s = (session *) malloc (sizeof (session));

  if (!s)
    {
      LL ();
      fprintf (stderr, "%s: session malloc() failed: %s (%i)\n", progname,
	       strerror (errno), errno);
      exit (EXIT_FAILURE);
    }
  else
    memset (s, '\0', sizeof (session));

  return s;
}

session *
find_session (const char *from, const char *to, time_t e, const char *randstr)
{
  session *scan;

  for (scan = s_top; scan; scan = scan->next)
    if (scan->e == e)
      if (!strcmp (from, scan->from))
	if (!strcmp (to, scan->to))
	  if (!strcmp (randstr, scan->randstr))
	    return scan;

  return (session *) 0;
}

session *
find_max (time_t e)
{
  session *scan;

  for (scan = s_top; scan; scan = scan->next)
    if (scan->e == e)
      return scan;

  return (session *) 0;
}

void
link_session (session * s)
{
  s->next = s_top;

  if (s_top)
    s_top->prev = s;
  s_top = s;
}

void
unlink_session (session * s)
{
  if (s->next)
    s->next->prev = s->prev;

  if (s->prev)
    s->prev->next = s->next;
  else
    s_top = s->next;

  s->next = (session *) 0;
  s->prev = (session *) 0;

  return;
}

void
show_sessions (void)
{
  session *scan;

  for (scan = s_top; scan; scan = scan->next)
    {
      fprintf (stderr, "%s\n", scan->from);
    }
}

session *
sort_list (session * list)
{
  session *p, *q, *e, *tail;
  int insize, nmerges, psize, qsize, i;

  if (!list)
    return (session *) 0;

  insize = 1;

  while (1)
    {
      p = list;
      list = (session *) 0;
      tail = (session *) 0;
      nmerges = 0;

      while (p)
	{
	  nmerges++;
	  q = p;
	  psize = 0;

	  for (i = 0; i < insize; i++)
	    {
	      psize++;
	      q = q->next;
	      if (!q)
		break;
	    }

	  qsize = insize;

	  while (psize > 0 || (qsize > 0 && q))
	    {
	      if (psize == 0)
		{
		  e = q;
		  q = q->next;
		  qsize--;
		}
	      else if (qsize == 0 || !q)
		{
		  e = p;
		  p = p->next;
		  psize--;
		}
	      else if (p->e < q->e)
		{
		  e = p;
		  p = p->next;
		  psize--;
		}
	      else
		{
		  e = q;
		  q = q->next;
		  qsize--;
		}

	      if (tail)
		tail->next = e;
	      else
		list = e;
	      tail = e;
	    }
	  p = q;
	}
      tail->next = (session *) 0;

      if (nmerges <= 1)
	return list;

      insize *= 2;
    }
}

void
count_dur (session * s)
{
  struct s_file s_time;
  char buffer[BUFSIZ];
  float t_time = 0;

  s_time.f = fopen (s->time.str, "r");

  if (s_time.f == (FILE *) 0)
    {
      LL ();
      fprintf (stderr, "%s: %s (%i)\n", s->time.str, strerror (errno), errno);
      exit (EXIT_FAILURE);
    }

  for (s_time.line = 1; fgets (buffer, BUFSIZ - 1, s_time.f); s_time.line++)
    {
      char time[BUFSIZ];
      float ftime = 0;
      int b = 0;

      if (sscanf (buffer, "%s %i\n", (char *) &time, &b) != 2)
	{			/* find the two numbers */
	  LL ();
	  fprintf (stderr,
		   "[error]: line %i: invalid timing information (!= %%f %%i\\n)\n",
		   s_time.line);
	  exit (EXIT_FAILURE);
	}

      if (sscanf (time, "%f", &ftime) != 1)
	{
	  LL ();
	  fprintf (stderr, "[error]: line %i: invalid float (!= %%f)",
		   s_time.line);
	  exit (EXIT_FAILURE);
	}

      t_time += ftime;
    }

  s->secs = (long) t_time;
  fclose (s_time.f);
}
